File: v1/start.js
const { WebBaseLoader } = require("langchain/document_loaders/web");
const { RecursiveCharacterTextSplitter } = require("langchain/text_splitter");
const { SklearnVectorStore } = require("langchain/vectorstores/sklearn");
const { OpenAIEmbeddings } = require("@langchain/openai");
const { ChatOllama } = require("@langchain/community/chat_models/ollama");
const { PromptTemplate } = require("@langchain/core/prompts");
const { StringOutputParser } = require("@langchain/core/output_parsers");
const { RunnableSequence } = require("@langchain/core/runnables");

// List of URLs
const urls = [
  "https://lilianweng.github.io/posts/2023-06-23-agent/",
  "https://lilianweng.github.io/posts/2023-03-15-prompt-engineering/",
  "https://lilianweng.github.io/posts/2023-10-25-adv-attack-llm/",
];

async function createRAGApplication() {
  // Load documents
  const loaders = urls.map(url => new WebBaseLoader(url));
  const docs = await Promise.all(loaders.map(loader => loader.load()));
  const docsList = docs.flat();

  // Initialize text splitter
  const textSplitter = RecursiveCharacterTextSplitter.fromDefaults({
    chunkSize: 250,
    chunkOverlap: 0,
  });

  // Split documents
  const docSplits = await textSplitter.splitDocuments(docsList);

  // Create vector store
  const vectorstore = await SklearnVectorStore.fromDocuments(
    docSplits,
    new OpenAIEmbeddings({ openAIApiKey: "api_key" })
  );
  const retriever = vectorstore.asRetriever(4);

  // Define prompt template
  const prompt = PromptTemplate.fromTemplate(`You are an assistant for question-answering tasks.
    Use the following documents to answer the question.
    If you don't know the answer, just say that you don't know.
    Use three sentences maximum and keep the answer concise:
    Question: {question}
    Documents: {documents}
    Answer:`);

  // Initialize LLM
  const llm = new ChatOllama({
    model: "llama3.2:1B",
    temperature: 0,
  });

  // Create chain
  const ragChain = RunnableSequence.from([
    {
      question: input => input.question,
      documents: async input => {
        const docs = await retriever.getRelevantDocuments(input.question);
        return docs.map(doc => doc.pageContent).join("\n");
      }
    },
    prompt,
    llm,
    new StringOutputParser()
  ]);

  return ragChain;
}

// Example usage
async function main() {
  const ragChain = await createRAGApplication();
  const question = "What kind of Server is Described in Section 3?";
  
  try {
    const answer = await ragChain.invoke({ question });
    console.log("Question:", question);
    console.log("Answer:", answer);
  } catch (error) {
    console.error("Error:", error);
  }
}

main(); 

File: start.py
from langchain_community.document_loaders import WebBaseLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
# List of URLs to load documents from
urls = [
    "<https://bitcoin.org/bitcoin.pdf>"
]

urlss = [
    "<https://lilianweng.github.io/posts/2023-06-23-agent/>",
    "<https://lilianweng.github.io/posts/2023-03-15-prompt-engineering/>",
    "<https://lilianweng.github.io/posts/2023-10-25-adv-attack-llm/>",
]
# Load documents from the URLs
docs = [WebBaseLoader(url).load() for url in urlss]
docs_list = [item for sublist in docs for item in sublist]


# Initialize a text splitter with specified chunk size and overlap
text_splitter = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
    chunk_size=250, chunk_overlap=0
)
# Split the documents into chunks
doc_splits = text_splitter.split_documents(docs_list)

from langchain_community.vectorstores import SKLearnVectorStore
from langchain_openai import OpenAIEmbeddings
# Create embeddings for documents and store them in a vector store
vectorstore = SKLearnVectorStore.from_documents(
    documents=doc_splits,
    embedding=OpenAIEmbeddings(openai_api_key="api_key"),
)
retriever = vectorstore.as_retriever(k=4)

from langchain_ollama import ChatOllama
from langchain.prompts import PromptTemplate
from langchain_core.output_parsers import StrOutputParser
# Define the prompt template for the LLM
prompt = PromptTemplate(
    template="""You are an assistant for question-answering tasks.
    Use the following documents to answer the question.
    If you don't know the answer, just say that you don't know.
    Use three sentences maximum and keep the answer concise:
    Question: {question}
    Documents: {documents}
    Answer:
    """,
    input_variables=["question", "documents"],
)

# Initialize the LLM with Llama 3.2:1B model
llm = ChatOllama(
    model="llama3.2:1B",
    temperature=0,
)

# Create a chain combining the prompt template and LLM
rag_chain = prompt | llm | StrOutputParser()

# Define the RAG application class
class RAGApplication:
    def __init__(self, retriever, rag_chain):
        self.retriever = retriever
        self.rag_chain = rag_chain
    def run(self, question):
        # Retrieve relevant documents
        documents = self.retriever.invoke(question)
        # Extract content from retrieved documents
        doc_texts = "\\n".join([doc.page_content for doc in documents])
        # Get the answer from the language model
        answer = self.rag_chain.invoke({"question": question, "documents": doc_texts})
        return answer
    
# Initialize the RAG application
rag_application = RAGApplication(retriever, rag_chain)
# Example usage
question = "What kind of Server is Described in Section 3?"
answer = rag_application.run(question)
print("Question:", question)
print("Answer:", answer)